tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
 	}
 }
 
+/*
+ * Check if the join fromList belongs to a INNER JOIN / LEFT JOIN
+ * on two tables. A INNER JOIN / LEFT JOIN on two tables has one JoinExpr
+ * entry in fromList.
+ */
+static bool
+is_inner_or_left_join_two_tables(List *fromList)
+{
+	if (!IsA(list_nth(fromList, 0), JoinExpr) || (list_length(fromList) != 1))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the join fromList belongs to an implicit join (e.g.,
+ * SELECT * FROM metric m1, metric_name m2 WHERE m1.id=m2.id). This type
+ * of join has two RANGETBLREF entries in the fromList.
+ */
+static bool
+is_implicit_join_two_tables(List *fromList)
```

Comment:
Is there anything special about the grammatically implicit join? It's just a particular kind of grammar for the usual inner join, right?

Reply:
They are indeed special. For left / right joins, you see a `JoinExpr` in the `fromList`. For these implicit joins, you have only `RangeTblRef` entries in the list (see the comment of the function `is_implicit_join_two_tables`).

Reply:
Hahah, I scroll down and here is an example of what I was talking about above! Just use reloptinfos :)

Reply:
In the most recent version of the PR, this function was removed and we don't access the `query->jointree` anymore.